### PR TITLE
Throw error message when MaxumimVersion is less than Version when using Import-Module cmdlet with -FullyQualifiedName parameter.

### LINKED ISSUE #7337: Get-Module accepts nonsensical combinations of ModuleVersion and MaximumVersion entries for -FullyQualifiedName
Steps to reproduce
------------------

Request a minimum version of 1.6.0 and - nonsensically - a maximum version of 1.5.0:

```powershell
Get-Module -ListAvailable -FullyQualifiedName @{ ModuleName='PowerShellGet'; ModuleVersion='1.6.0'; MaximumVersion='1.5.0' }
```

Expected behavior
-----------------

I think a statement-terminating error indicating that `MaximumVersion`  must be greater or equal to `ModuleVersion` (which uses _minimum_ logic) would be more helpful than the current behavior.

If feasible, this validation should be tied to fully qualified module names themselves, wherever they are accepted.

Actual behavior
---------------

The argument is quietly accepted and therefore results in no output (because, by definition, no modules match).

Environment data
----------------

```powershell
PowerShell Core v6.1.0-preview.3 on macOS 10.13.6
PowerShell Core v6.1.0-preview.3 on Ubuntu 16.04.4 LTS
PowerShell Core v6.1.0-preview.3 on Microsoft Windows 10 Pro (64-bit; Version 1803, OS Build: 17134.165)
\Windows PowerShell v5.1.17134.165 on Microsoft Windows 10 Pro (64-bit; Version 1803, OS Build: 17134.165)
```

### PR DESCRIPTION
**Pull request by sethvs**
## PR Summary

Fix #7337
 
Also fixing following error message to include 'MaximumVersion' in the list of 'valid members':
```
  <data name="InvalidModuleSpecificationMember" xml:space="preserve">
    <value>The hashtable describing a module contains one or more members that are not valid.  The valid members are ({0}). Remove the members that are not valid ({1}), then try again.</value>
  </data>
```
Concerning if the change is breaking, I would refer it to Unlikely Grey Area.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [ ] Not Applicable or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

### CODE REVIEW

File: src/System.Management.Automation/engine/Modules/ModuleSpecification.cs [hunk 1]
```diff
@@ -136,7 +136,7 @@ internal static Exception ModuleSpecificationInitHelper(ModuleSpecification modu
             string message;
             if (badKeys.Length != 0)
             {
-                message = StringUtil.Format(Modules.InvalidModuleSpecificationMember, "ModuleName, ModuleVersion, RequiredVersion, GUID", badKeys);
+                message = StringUtil.Format(Modules.InvalidModuleSpecificationMember, "ModuleName, ModuleVersion, MaximumVersion, RequiredVersion, GUID", badKeys);
```
Comment (anmenaga) [inline src/System.Management.Automation/engine/Modules/ModuleSpecification.cs]: Please add to PR Summary:
```
Also fixing following error message to include 'MaximumVersion' in the list of 'valid members':
  <data name="InvalidModuleSpecificationMember" xml:space="preserve">
    <value>The hashtable describing a module contains one or more members that are not valid.  The valid members are ({0}). Remove the members that are not valid ({1}), then try again.</value>
  </data>
```

**Comment by anmenaga** [comment]
@daxian-dbw @SteveL-MSFT please evaluate if this technically breaking (however unlikely) change has to go to PS Committee for review.

File: src/System.Management.Automation/engine/Modules/ModuleSpecification.cs (hunk 1 above)
Comment (sethvs) [reply src/System.Management.Automation/engine/Modules/ModuleSpecification.cs]: Done.

**Comment by SteveL-MSFT** [comment]
@PowerShell/powershell-committee agrees this is a breaking change.  For example using this in a pipeline would terminate the pipeline now whereas a $null being returned would not execute the rest of the pipeline.  We agree that this should have been an error and would accept returning a non-terminating error and not a terminating one.

**Comment by sethvs** [comment]
Can you guide me a little bit?
How can I use WriteError in there?
https://github.com/PowerShell/PowerShell/blob/321537c6e52b24e9187afd267bb0b18b242bbed3/src/System.Management.Automation/engine/Modules/ModuleSpecification.cs#L167-L172

Or is there another way to return non-terminating error?

**Comment by SteveL-MSFT** [comment]
@sethvs I took a look at the code.  Since the validation is in the ModuleSpecification class that should continue to return an exception that gets thrown.  You'll probably need to update GetModuleCommand.cs to handle this specific case.

File: src/System.Management.Automation/engine/Modules/ModuleSpecification.cs (hunk 1 above)
Comment (rjmholt) [reply src/System.Management.Automation/engine/Modules/ModuleSpecification.cs]: Ideally this should use `$"{nameof(...)}"` for the bad members

File: src/System.Management.Automation/resources/Modules.resx [hunk 2]
```diff
@@ -165,6 +165,9 @@
   <data name="InvalidModuleSpecificationMember" xml:space="preserve">
     <value>The hashtable describing a module contains one or more members that are not valid.  The valid members are ({0}). Remove the members that are not valid ({1}), then try again.</value>
   </data>
+  <data name="ModuleSpecificationMemberIsLessThanOther" xml:space="preserve">
```
Comment (rjmholt) [inline src/System.Management.Automation/resources/Modules.resx]: The error message should ideally be slightly more specific. Something like:
> The hashtable describing a module cannot have its ModuleVersion ({0}) greater than its MaximumVersion ({1}).

File: test/powershell/Modules/Microsoft.PowerShell.Core/Import-Module.Tests.ps1 [hunk 3]
```diff
@@ -56,6 +56,10 @@ Describe "Import-Module" -Tags "CI" {
         Import-Module TestModule -RequiredVersion 1.1
         (Get-Module TestModule).Version | Should -BeIn "1.1"
     }
+
+    It "should throw if 'MaximumVersion' is less than 'Version' when using -FullyQualifiedName parameter" {
```
Comment (rjmholt) [inline test/powershell/Modules/Microsoft.PowerShell.Core/Import-Module.Tests.ps1]: In addition to this test, we should add a more specific one using the `[ModuleSpecification]` constructor.

You should find those tests [here](https://github.com/PowerShell/PowerShell/blob/master/test/powershell/engine/Module/ModuleSpecification.Tests.ps1) (I know that's not an obvious place at all).

Note to self: once https://github.com/PowerShell/PowerShell/pull/7499 is merged, this test will be included in that.

**Review by rjmholt — COMMENTED** [review]
Apologies, I should have reviewed this sooner. Looks like a good change! Just left a couple of minor comments.

Thanks for your contribution @sethvs!

File: test/powershell/Modules/Microsoft.PowerShell.Core/Import-Module.Tests.ps1 (hunk 3 above)
Comment (rjmholt) [reply test/powershell/Modules/Microsoft.PowerShell.Core/Import-Module.Tests.ps1]: Let me know if you need any help on creating that test btw, I'm always happy to

**Comment by rjmholt** [comment]
@SteveL-MSFT is it possible for a parameter binding exception to be non-terminating? The error when called with `Get-Module` will occur when the hashtable is cast to a `ModuleSpecification`; execution will never reach `Get-Module`.

**Comment by rjmholt** [comment]
Also for reference, here is `Import-Module`'s version check:
https://github.com/PowerShell/PowerShell/blob/c94fc31f9767fe5bc835ee4921896b6331eaad4d/src/System.Management.Automation/engine/Modules/ImportModuleCommand.cs#L1686-L1690

I think that's a terminating error, no?

File: src/System.Management.Automation/engine/Modules/ModuleSpecification.cs (hunk 1 above)
Comment (sethvs) [reply src/System.Management.Automation/engine/Modules/ModuleSpecification.cs]: nameof(badKeys) returns 'badKeys', but we need its actual value.
Or am I missing something?

File: src/System.Management.Automation/resources/Modules.resx (hunk 2 above)
Comment (sethvs) [reply src/System.Management.Automation/resources/Modules.resx]: Done.

File: test/powershell/Modules/Microsoft.PowerShell.Core/Import-Module.Tests.ps1 (hunk 3 above)
Comment (sethvs) [reply test/powershell/Modules/Microsoft.PowerShell.Core/Import-Module.Tests.ps1]: Done.

**Comment by sethvs** [comment]
Is this OK with VSTS: PowerShell-CI-macos?

**Review by anmenaga — APPROVED** [review]

**Comment by sethvs** [comment]
Close and reopen to restart tests.

**Comment by SteveL-MSFT** [comment]
@PowerShell/powershell-committee re-reviewed this.  The conclusion is that well-formed filters should return nothing if the result set is nothing and not return an error.  In this case, the filter is well-formed, although invalid.  Since this is also a breaking change, our recommendation is to not take this change as it's inconsistent with the desired filtering behavior and introduces a breaking change.

**Pull request closed without merging**